contracts/TokenLocking.sol

```diff
@@ -95,39 +98,49 @@ contract TokenLocking is TokenLockingStorage, DSMath { // ignore-swc-123
     userLocks[_token][msg.sender].lockCount = _lockId;
   }
 
-  uint256 constant UINT192_MAX = 2**192 - 1; // Used for updating the deposit timestamp
-
-  function deposit(address _token, uint256 _amount) public
-  tokenNotLocked(_token)
+  function deposit(address _token, uint256 _amount, bool _force) public
+  tokenNotLocked(_token, _force)
   {
-    require(_amount > 0, "colony-token-locking-invalid-amount");
+    Lock storage lock = userLocks[_token][msg.sender];
     require(ERC20Extended(_token).transferFrom(msg.sender, address(this), _amount), "colony-token-locking-transfer-failed"); // ignore-swc-123
 
-    Lock storage lock = userLocks[_token][msg.sender];
+    lock.timestamp = getNewTimestamp(lock.balance, _amount, lock.timestamp, now);
+    lock.balance = add(lock.balance, _amount);
 
-    uint256 prevWeight = lock.balance;
-    uint256 currWeight = _amount;
+    emit UserTokenDeposited(_token, msg.sender, lock.balance, lock.timestamp);
+  }
 
-    // Needed to prevent overflows in the timestamp calculation
-    while ((prevWeight >= UINT192_MAX) || (currWeight >= UINT192_MAX)) {
-      prevWeight /= 2;
-      currWeight /= 2;
-    }
+  function claim(address _token, bool _force) public
+  tokenNotLocked(_token, _force)
+  {
+    Lock storage lock = userLocks[_token][msg.sender];
 
-    uint256 newAmount = add(lock.balance, _amount);
-    uint256 newTimestamp = add(mul(prevWeight, lock.timestamp), mul(currWeight, now)) / add(prevWeight, currWeight);
-    userLocks[_token][msg.sender] = Lock(totalLockCount[_token], newAmount, newTimestamp);
+    lock.timestamp = getNewTimestamp(lock.balance, lock.pendingBalance, lock.timestamp, now);
+    lock.balance = add(lock.balance, lock.pendingBalance);
+    lock.pendingBalance = 0;
 
-    emit UserTokenDeposited(_token, msg.sender, newAmount, newTimestamp);
+    emit UserTokenClaimed(_token, msg.sender, lock.balance, lock.timestamp);
   }
 
-  function withdraw(address _token, uint256 _amount) public
-  tokenNotLocked(_token)
+  function transfer(address _token, uint256 _amount, address _recipient, bool _force) public
```

Comment:
Yes, but I think it's justified in that only I can call deposit for myself, so I have more control over it -- I know if there's an active lock or not, etc. Whereas for transfer, someone else might call it on my behalf, so I don't want a possibility of the transaction failing.

I had considered adding logic which checks for active locks and does conditional behavior, but ultimately wanted the simplicity of having only one behavior. I can add the branching logic if you like though, it wouldn't be particularly difficult.

Reply:
Yeah, I agree with that - if I want to `transfer` I definitely should be able to, regardless of the state the destination address is in. I think the branching logic, while making the code in the contracts slightly more complicated, will make the user interface / experience far simpler.

Reply:
Ok, I can update.

Reply:
The only thing about this that's a little iffy is that it lets someone other than the user influence the deposit timestamp. I don't think it's really a problem, but I wonder if there's a way someone could transfer me tokens to push my deposit timestamp back, which would affect my ability to submit reputation root hashes.

Reply:
Ok, here's my position on this:

Yes it is a possible attack, and it's impossible for us to prove anything about the safety or cost of such an attack since we can't anticipate all the future ways colonies might use the deposit timestamp. But since the attack involves literally giving someone money, we can advise colonies to design their mechanisms in such a way that the value of an early timestamp is never more than what it would cost to push that timestamp forward in time, or in other words, that no one can ever lose more money due to a more recent timestamp than it would cost an attacker to move the timestamp. In that scenario, the attack is never worthwhile.